include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,346 @@
+                            m_partition{partition},
+                            m_local_index{std::move(local_index)},
+                            m_levels{levels} {}
+                        // less safe but maybe preferable
+                        // template<typename V>
+                        // iteration_space(const int partition, V&& remote_index) noexcept :
+                        //     m_partition{partition},
+                        //     m_remote_index{std::forward<V>(remote_index)} {}
+                        iteration_space(const int partition, const index_type first, const index_type last, const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{},
+                            m_levels{levels} {
+                            m_local_index.resize(static_cast<std::size_t>(last - first + 1));
+                            for (index_type idx = first; idx <= last; ++idx) {
+                                m_local_index[static_cast<std::size_t>(idx)] = idx;
+                            }
+                        }
+
+                        // member functions
+                        int partition() const noexcept { return m_partition; }
+                        // std::vector<index_type, gridtools::allocator::unified_memory_allocator<index_type>>& local_index() noexcept { return m_local_index; }
+                        const std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>>& local_index() const noexcept { return m_local_index; }
+                        std::size_t levels() const noexcept { return m_levels; }
+                        std::size_t size() const noexcept { return m_local_index.size() * m_levels; }
+
+                        // print
+                        /** @brief print */
+                        template<class CharT, class Traits>
+                        friend std::basic_ostream<CharT, Traits>& operator << (std::basic_ostream<CharT, Traits>& os, const iteration_space& is) {
+                            os << "size = " << is.size() << ";\n"
+                               << "# levels = " << is.levels() << ";\n"
+                               << "partition = " << is.partition() << ";\n"
+                               << "local indexes: [ ";
+                            for (auto idx : is.local_index()) { os << idx << " "; }
+                            os << "]\n";
+                            return os;
+                        }
+
+                };
+
+                using iteration_space_pair = iteration_space;
+                using index_container_type = std::vector<iteration_space_pair>;
+
+                /** @brief extended domain id, including rank and tag information
+                 * WARN: domain id temporarily set equal to rank, differently from structured pattern case*/
+                struct extended_domain_id_type {
```

Comment:
Maybe we should put these classes outside the scope of the class, maybe in a namespace, because reading the code is not easy now.

Reply:
Yep, I agree. Only thing is that I think we should be consistent in `structured/pattern.hpp` as well, so if you say I can refactor that one as well.